### PR TITLE
reduce build time 20%

### PR DESCRIPTION
**Pull request by shibukawa**
I squashed commits and do force push, but github doesn't reopen the ticket I recreated pull request.

-------

CMake tries to build same source file multiple times for each target. Sharing source files(.cpp) increases build time instead of sharing static libraries.

Building time becomes 50s to 40s on my computer :)

Before
=========

```
[shibukawa.yoshiki]$ time make                                                                                                                                                                                                    [master]
-- No build type selected, default to Release
-- Building with -std=c++11
-- Building with -msse2
-- Building with -mfpmath=sse
-- Building with -Wall
-- Building with -Werror
-- Building with -Wextra
-- Building with -Wno-unused-parameter
-- Building with -fno-omit-frame-pointer
-- Building with -O2
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/shibukawa.yoshiki/Library/Caches/qtpm/binaryen
[  2%] Building CXX object CMakeFiles/support.dir/src/support/bits.cpp.o
[  4%] Building CXX object CMakeFiles/support.dir/src/support/colors.cpp.o
[  6%] Building CXX object CMakeFiles/support.dir/src/support/command-line.cpp.o
[  8%] Building CXX object CMakeFiles/support.dir/src/support/file.cpp.o
[ 10%] Building CXX object CMakeFiles/support.dir/src/support/safe_integer.cpp.o
[ 12%] Linking CXX static library lib/libsupport.a
[ 12%] Built target support
[ 14%] Building CXX object CMakeFiles/asm2wasm.dir/src/asm2wasm-main.cpp.o
[ 16%] Building CXX object CMakeFiles/asm2wasm.dir/src/pass.cpp.o
[ 18%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/MergeBlocks.cpp.o
[ 20%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/PostEmscripten.cpp.o
[ 22%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/Print.cpp.o
[ 25%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/RemoveUnusedBrs.cpp.o
[ 27%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/RemoveUnusedNames.cpp.o
[ 29%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/SimplifyLocals.cpp.o
[ 31%] Building CXX object CMakeFiles/asm2wasm.dir/src/passes/ReorderLocals.cpp.o
[ 33%] Building CXX object CMakeFiles/asm2wasm.dir/src/emscripten-optimizer/parser.cpp.o
[ 35%] Building CXX object CMakeFiles/asm2wasm.dir/src/emscripten-optimizer/simple_ast.cpp.o
[ 37%] Building CXX object CMakeFiles/asm2wasm.dir/src/emscripten-optimizer/optimizer-shared.cpp.o
[ 39%] Linking CXX executable bin/asm2wasm
[ 39%] Built target asm2wasm
[ 41%] Building CXX object CMakeFiles/binaryen-shell.dir/src/binaryen-shell.cpp.o
[ 43%] Building CXX object CMakeFiles/binaryen-shell.dir/src/pass.cpp.o
[ 45%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/LowerIfElse.cpp.o
[ 47%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/MergeBlocks.cpp.o
[ 50%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/NameManager.cpp.o
[ 52%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/PostEmscripten.cpp.o
[ 54%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/Print.cpp.o
[ 56%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/RemoveImports.cpp.o
[ 58%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/RemoveUnusedBrs.cpp.o
[ 60%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/RemoveUnusedNames.cpp.o
[ 62%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/SimplifyLocals.cpp.o
[ 64%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/ReorderLocals.cpp.o
[ 66%] Building CXX object CMakeFiles/binaryen-shell.dir/src/passes/Metrics.cpp.o
[ 68%] Linking CXX executable bin/binaryen-shell
[ 68%] Built target binaryen-shell
[ 70%] Building CXX object CMakeFiles/s2wasm.dir/src/pass.cpp.o
[ 72%] Building CXX object CMakeFiles/s2wasm.dir/src/passes/Print.cpp.o
[ 75%] Building CXX object CMakeFiles/s2wasm.dir/src/s2wasm-main.cpp.o
[ 77%] Linking CXX executable bin/s2wasm
[ 77%] Built target s2wasm
[ 79%] Building CXX object CMakeFiles/wasm-as.dir/src/wasm-as.cpp.o
[ 81%] Linking CXX executable bin/wasm-as
[ 81%] Built target wasm-as
[ 83%] Building CXX object CMakeFiles/wasm-dis.dir/src/pass.cpp.o
[ 85%] Building CXX object CMakeFiles/wasm-dis.dir/src/passes/Print.cpp.o
[ 87%] Building CXX object CMakeFiles/wasm-dis.dir/src/wasm-dis.cpp.o
[ 89%] Linking CXX executable bin/wasm-dis
[ 89%] Built target wasm-dis
[ 91%] Building CXX object CMakeFiles/wasm2asm.dir/src/wasm2asm-main.cpp.o
[ 93%] Building CXX object CMakeFiles/wasm2asm.dir/src/emscripten-optimizer/parser.cpp.o
[ 95%] Building CXX object CMakeFiles/wasm2asm.dir/src/emscripten-optimizer/simple_ast.cpp.o
[ 97%] Building CXX object CMakeFiles/wasm2asm.dir/src/emscripten-optimizer/optimizer-shared.cpp.o
[100%] Linking CXX executable bin/wasm2asm
[100%] Built target wasm2asm
make  47.60s user 2.67s system 99% cpu 50.739 total
```

After
=========

```
[shibukawa.yoshiki]$ time make                                                                                                                                                                                                    [master]
Scanning dependencies of target optimizer
[  2%] Building CXX object CMakeFiles/optimizer.dir/src/emscripten-optimizer/parser.cpp.o
[  5%] Building CXX object CMakeFiles/optimizer.dir/src/emscripten-optimizer/simple_ast.cpp.o
[  8%] Building CXX object CMakeFiles/optimizer.dir/src/emscripten-optimizer/optimizer-shared.cpp.o
[ 11%] Linking CXX static library lib/liboptimizer.a
[ 11%] Built target optimizer
[ 14%] Building CXX object CMakeFiles/support.dir/src/support/bits.cpp.o
[ 17%] Building CXX object CMakeFiles/support.dir/src/support/colors.cpp.o
[ 20%] Building CXX object CMakeFiles/support.dir/src/support/command-line.cpp.o
[ 22%] Building CXX object CMakeFiles/support.dir/src/support/file.cpp.o
[ 25%] Building CXX object CMakeFiles/support.dir/src/support/safe_integer.cpp.o
[ 28%] Linking CXX static library lib/libsupport.a
[ 28%] Built target support
Scanning dependencies of target pass
[ 31%] Building CXX object CMakeFiles/pass.dir/src/pass.cpp.o
[ 34%] Building CXX object CMakeFiles/pass.dir/src/passes/LowerIfElse.cpp.o
[ 37%] Building CXX object CMakeFiles/pass.dir/src/passes/MergeBlocks.cpp.o
[ 40%] Building CXX object CMakeFiles/pass.dir/src/passes/NameManager.cpp.o
[ 42%] Building CXX object CMakeFiles/pass.dir/src/passes/PostEmscripten.cpp.o
[ 45%] Building CXX object CMakeFiles/pass.dir/src/passes/Print.cpp.o
[ 48%] Building CXX object CMakeFiles/pass.dir/src/passes/RemoveImports.cpp.o
[ 51%] Building CXX object CMakeFiles/pass.dir/src/passes/RemoveUnusedBrs.cpp.o
[ 54%] Building CXX object CMakeFiles/pass.dir/src/passes/RemoveUnusedNames.cpp.o
[ 57%] Building CXX object CMakeFiles/pass.dir/src/passes/SimplifyLocals.cpp.o
[ 60%] Building CXX object CMakeFiles/pass.dir/src/passes/ReorderLocals.cpp.o
[ 62%] Building CXX object CMakeFiles/pass.dir/src/passes/Metrics.cpp.o
[ 65%] Linking CXX static library lib/libpass.a
[ 65%] Built target pass
Scanning dependencies of target asm2wasm
[ 68%] Building CXX object CMakeFiles/asm2wasm.dir/src/asm2wasm-main.cpp.o
[ 71%] Linking CXX executable bin/asm2wasm
[ 71%] Built target asm2wasm
Scanning dependencies of target binaryen-shell
[ 74%] Building CXX object CMakeFiles/binaryen-shell.dir/src/binaryen-shell.cpp.o
[ 77%] Linking CXX executable bin/binaryen-shell
[ 77%] Built target binaryen-shell
Scanning dependencies of target s2wasm
[ 80%] Building CXX object CMakeFiles/s2wasm.dir/src/s2wasm-main.cpp.o
[ 82%] Linking CXX executable bin/s2wasm
[ 82%] Built target s2wasm
[ 85%] Building CXX object CMakeFiles/wasm-as.dir/src/wasm-as.cpp.o
[ 88%] Linking CXX executable bin/wasm-as
[ 88%] Built target wasm-as
Scanning dependencies of target wasm-dis
[ 91%] Building CXX object CMakeFiles/wasm-dis.dir/src/wasm-dis.cpp.o
[ 94%] Linking CXX executable bin/wasm-dis
[ 94%] Built target wasm-dis
Scanning dependencies of target wasm2asm
[ 97%] Building CXX object CMakeFiles/wasm2asm.dir/src/wasm2asm-main.cpp.o
[100%] Linking CXX executable bin/wasm2asm
[100%] Built target wasm2asm
make  35.94s user 2.08s system 98% cpu 38.415 total
```

### CODE REVIEW

**Comment by kripken** [comment]
Looks like this fails on the bots. It can't find a pass, which implies that maybe you aren't linking in all the files under `src/passes`/ into the binaryen shell, or another tool?

**Comment by shibukawa** [comment]
I read passes code and understood. Linking code has side-effects because source codes in pases folder register theirselves. So my initial approach was wrong (it links all codes as a static library).

Now libpass.a contains minimum codes. The result becomes 50s -> 43s.

**Comment by kripken** [comment]
Thanks, looks good to me know. But might be good if someone that knows cmake better than me (which is pretty much everyone :) ) could take a look too.

Also, please join the [community group](https://www.w3.org/community/webassembly/participants) before we merge, if you haven't already.

File: CMakeLists.txt [hunk 1]
```diff
@@ -78,14 +78,26 @@ SET(support_SOURCES
...
+  src/emscripten-optimizer/optimizer-shared.cpp
+)
+ADD_LIBRARY(optimizer STATIC ${optimizer_SOURCES})
+
+
+SET(pass_SOURCES
+   src/pass.cpp
+   src/passes/Print.cpp
+ )
+ADD_LIBRARY(pass STATIC ${pass_SOURCES})
```
Comment (jfbastien) [inline CMakeLists.txt]: This is an odd way to organize things. In general I'd like static libs to contain everything under a folder, and nothing else (i.e. folder hierarchy reflects library hierarchy). It makes it way easier to understand how things work.

pass.cpp should move under passes/, and parres/ contains a bunch of passes which should be in passes/.

The "right" way to keep this clean is probably to use cmake's add_subdirectory from the root cmake file, and then have each subdir create whatever it wants (static libraries most likely).

**Comment by jfbastien** [comment]
I'm OK if https://github.com/WebAssembly/binaryen/pull/302#discussion_r58278110 is done separately or here. I don't want to block this change on it, but I think it'll keep things sane.

**Comment by kripken** [comment]
How about if we have a single `binaryen.so` shared library, that all the tools would be linked with dynamically? That would be good for other projects using binaryen as well, they would just link with it similarly.

**Comment by jfbastien** [comment]
> How about if we have a single binaryen.so shared library, that all the tools would be linked with dynamically? That would be good for other projects using binaryen as well, they would just link with it similarly.

I think it's better to do small static libraries where possible: it keeps code sane, prevents introduction of odd dependencies, and keeps fast build time. Not that dylibs are bad either, it just seems to be a slightly unrelated thing: the end tools build faster, but you still need to rebuild the `.so`.

**Comment by kripken** [comment]
Small static libs internally sounds ok, but are you fine with also having a single `.so` for external projects? e.g. the program in examples/, I'd like it to just link to `libbinaryen` and not to multiple static libs.

**Comment by jfbastien** [comment]
> Small static libs internally sounds ok, but are you fine with also having a single `.so` for external projects? e.g. the program in examples/, I'd like it to just link to `libbinaryen` and not to multiple static libs.

Yeah that sounds good, I'd tackle it after the static libs though.

**Comment by kripken** [comment]
Sounds reasonable.

**Comment by shibukawa** [comment]
I updated my PR to use ADD_SUBDIRECTORY.

**Comment by jfbastien** [comment]
Awesome, I think that's a pretty good improvement on its own. We can probably do more in that vein afterwards.

**Comment by kripken** [comment]
Looks nice, thanks. Please just rebase so this can be merged, and also I think you need to join the [WebAssembly w3c group](https://www.w3.org/community/webassembly/participants).

**Comment by dschuff** [comment]
@shibukawa Do you plan on rebasing this? If not, do you mind if I do that and land it? I have another refactoring that would be much nicer if this were in with it.

**Comment by shibukawa** [comment]
If you are making better solution, feel free to discard my commit.

**Comment by kripken** [comment]
Ok, another solution has landed for this, so I think this can be closed.

**Pull request closed without merging**